doozerlib/metadata.py

```diff
@@ -190,37 +294,109 @@ def cgit_url(self, filename: str, commit_hash: Optional[str] = None, branch: Opt
+                         build's release field. Pattern must match release timestamp and components
+                         like p? and git commit (up to, but not including ".assembly.<name>" release
+                         component). e.g. "*.git.<commit>.*   or '*.p1.*'
+        :param build_state: 0=BUILDING, 1=COMPLETE, 2=DELETED, 3=FAILED, 4=CANCELED
+        :return: Returns the most recent build object from koji for this package & assembly.
+                 Example https://gist.github.com/jupierce/57e99b80572336e8652df3c6be7bf664
         """
-        component_name = self.get_component_name()
+        package_name = self.get_component_name()
+        builds = []
+
         with self.runtime.pooled_koji_client_session() as koji_api:
-            builds = koji_api.getLatestBuilds(self.candidate_brew_tag(), package=component_name)
-            if not builds:
-                if default != -1:
-                    self.logger.warning("No builds detected for using tag: %s" % (self.candidate_brew_tag()))
-                    return default
-                raise IOError("No builds detected for %s using tag: %s" % (self.qualified_name, self.candidate_brew_tag()))
-            return builds[0]
-
-    def get_latest_build_info(self, default=-1):
+            package_info = koji_api.getPackage(package_name)  # e.g. {'id': 66873, 'name': 'atomic-openshift-descheduler-container'}
+            if not package_info:
+                raise IOError(f'No brew package is defined for {package_name}')
+            package_id = package_info['id']  # we could just constrain package name using pattern glob, but providing package ID # should be a much more efficient DB query.
+            # listBuilds returns all builds for the package; We need to limit the query to the builds
+            # relevant for our major/minor.
+
+            # RPMs do not have a 'v' in front of their version; images do.
+            if self.meta_type == 'image':
+                ver_prefix = 'v'  # openshift-enterprise-console-container-v4.7.0-202106032231.p0.git.d9f4379
+            else:
+                ver_prefix = ''  # openshift-clients-4.7.0-202106032231.p0.git.e29b355.el8
+
+            pattern_prefix = f'{package_name}-{ver_prefix}{self.branch_major_minor()}.'
+
+            if assembly is None:
+                assembly = self.runtime.assembly
+
+            def latest_build_list(pattern_suffix):
+                # We need '*' after pattern_suffix for RPMs, which have an extra .el7/.el8 on their
+                # release field, normally.
+                builds = koji_api.listBuilds(packageID=package_id,
+                                             state=None if build_state is None else build_state.value,
+                                             pattern=f'{pattern_prefix}{extra_pattern}{pattern_suffix}*',
+                                             queryOpts={'limit': 1, 'order': '-creation_event_id'})
```

Comment:
For rpms, we are building for both -el7 and -el8 tags. This call could return a build for either el7 or el8 (depending on the timing).

Reply:
Thanks, @vfreex! I had forgotten about multi-target RPMs. Added `el_target` to allow the user to specify intent. I'm then using that in `needs_rebuild` to check all targets.